File: TrainBlink/android/app/src/main/java/com/trainblink/app/managers/ChatManager.kt
package com.trainblink.app.managers

import android.content.Context
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.trainblink.app.TrainBlinkApplication
import com.trainblink.app.models.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.util.Date

/**
 * Chat Manager
 * Feature 5: Chat messaging system
 * Android equivalent of iOS ChatManager with SharedPreferences persistence
 */
class ChatManager(private val context: Context) {

    private val prefs = context.getSharedPreferences("trainblink_chat", Context.MODE_PRIVATE)
    private val gson = Gson()
    private val analyticsManager get() = TrainBlinkApplication.instance.analyticsManager
    private val blockingManager get() = TrainBlinkApplication.instance.blockingManager
    private val encounterTrackingManager get() = TrainBlinkApplication.instance.encounterTrackingManager
    private val ephemeralMessageManager get() = TrainBlinkApplication.instance.ephemeralMessageManager

    private val _chatRooms = MutableStateFlow<List<ChatRoom>>(emptyList())
    val chatRooms: StateFlow<List<ChatRoom>> = _chatRooms.asStateFlow()

    private var myPeerId: String = ""
    private var nearbyConnectionsManager: NearbyConnectionsManager? = null

    init {
        loadChatRooms()
        setupEphemeralMessageHandler()
        println("💬 ChatManager initialized - ${_chatRooms.value.size} chat rooms")
    }

    /**
     * Set my peer ID
     */
    fun setMyPeerId(peerId: String) {
        myPeerId = peerId
        println("💬 My peer ID: $peerId")
    }

    /**
     * Set Nearby Connections Manager for message delivery
     */
    fun setNearbyConnectionsManager(manager: NearbyConnectionsManager) {
        nearbyConnectionsManager = manager
        println("💬 Connected to NearbyConnectionsManager")
    }

    /**
     * Setup ephemeral message expiration handler
     */
    private fun setupEphemeralMessageHandler() {
        ephemeralMessageManager.onMessageExpired = { messageId ->
            deleteMessage(messageId)
        }
    }

    /**
     * Send a message
     */
    fun sendMessage(
        text: String,
        receiverId: String,
        isEphemeral: Boolean = false,
        expirationSeconds: Int? = null
    ): ChatMessage? {
        // Check if receiver is blocked
        if (blockingManager.isBlocked(receiverId)) {
            println("⚠️ Cannot send message to blocked peer: $receiverId")
            return null
        }

        // Create message
        val expiresAt = if (isEphemeral && expirationSeconds != null) {
            Date(System.currentTimeMillis() + expirationSeconds * 1000L)
        } else null

        val message = ChatMessage(
            text = text,
            senderId = myPeerId,
            receiverId = receiverId,
            deliveryStatus = MessageDeliveryStatus.SENDING,
            isEphemeral = isEphemeral,
            expiresAt = expiresAt
        )

        // Add to chat room
        addMessageToChatRoom(message)

        // Track ephemeral message
        if (isEphemeral) {
            ephemeralMessageManager.trackMessage(message)
        }

        // Send via Nearby Connections
        val success = sendMessageViaNearby(message)

        // Update delivery status
        val updatedMessage = if (success) {
            message.markAsDelivered()
        } else {
            message.copy(deliveryStatus = MessageDeliveryStatus.FAILED)
        }
        updateMessage(updatedMessage)

        // Record encounter
        val peerDisplayName = getChatRoom(receiverId)?.peerDisplayName ?: "Unknown"
        val peer = Peer(
            id = receiverId,
            displayName = peerDisplayName,
            connectionState = PeerConnectionState.CONNECTED,
            lastSeenAt = Date()
        )
        encounterTrackingManager.recordEncounter(
            peer = peer,
            interactionType = InteractionType.CHAT
        )

        // Log analytics
        analyticsManager.logMessageSent(
            peerId = receiverId,
            isEphemeral = isEphemeral,
            messageLength = text.length
        )

        println("💬 Message sent: ${message.id} (${if (isEphemeral) "ephemeral" else "normal"})")
        return updatedMessage
    }

    /**
     * Receive a message
     */
    fun handleReceivedMessage(data: ByteArray, from: String) {
        try {
            val message = gson.fromJson(String(data), ChatMessage::class.java)

            // Check if sender is blocked
            if (blockingManager.isBlocked(message.senderId)) {
                println("🚫 Blocked message from: ${message.senderId}")
                return
            }

            // Add to chat room
            addMessageToChatRoom(message)

            // Track ephemeral message
            if (message.isEphemeral) {
                ephemeralMessageManager.trackMessage(message)
            }

            // Record encounter
            val peer = Peer(
                id = message.senderId,
                displayName = message.senderDisplayName ?: "Unknown",
                connectionState = PeerConnectionState.CONNECTED,
                lastSeenAt = Date()
            )
            encounterTrackingManager.recordEncounter(
                peer = peer,
                interactionType = InteractionType.CHAT
            )

            // Log analytics
            analyticsManager.logMessageReceived(
                peerId = message.senderId,
                isEphemeral = message.isEphemeral,
                messageLength = message.text.length
            )

            println("💬 Message received: ${message.id} from ${message.senderId}")
        } catch (e: Exception) {
            println("❌ Failed to handle received message: ${e.message}")
        }
    }

    /**
     * Get or create chat room for a peer
     */
    fun getOrCreateChatRoom(peerId: String, peerDisplayName: String): ChatRoom {
        val existing = getChatRoom(peerId)
        if (existing != null) {
            return existing
        }

        val newRoom = ChatRoom(
            peerId = peerId,
            peerDisplayName = peerDisplayName
        )
        _chatRooms.value = _chatRooms.value + newRoom
        saveChatRooms()

        println("💬 Created chat room: $peerDisplayName")
        return newRoom
    }

    /**
     * Get chat room for a peer
     */
    fun getChatRoom(peerId: String): ChatRoom? {
        return _chatRooms.value.find { it.peerId == peerId }
    }

    /**
     * Get all messages for a peer
     */
    fun getMessages(peerId: String): List<ChatMessage> {
        return getChatRoom(peerId)?.messages ?: emptyList()
    }

    /**
     * Get sorted chat rooms (by last message)
     */
    fun getSortedChatRooms(): List<ChatRoom> {
        return _chatRooms.value.sortedByDescending {
            it.lastMessage?.timestamp?.time ?: 0
        }
    }

    /**
     * Get active chat rooms (with messages)
     */
    fun getActiveChatRooms(): List<ChatRoom> {
        return _chatRooms.value.filter { it.messages.isNotEmpty() }
    }

    /**
     * Mark chat room as read
     */
    fun markChatRoomAsRead(peerId: String) {
        val room = getChatRoom(peerId) ?: return
        val updatedRoom = room.markAsRead()
        updateChatRoom(updatedRoom)
        println("💬 Marked chat room as read: $peerId")
    }

    /**
     * Delete a message
     */
    fun deleteMessage(messageId: String) {
        _chatRooms.value = _chatRooms.value.map { room ->
            val updatedMessages = room.messages.filter { it.id != messageId }
            if (updatedMessages.size != room.messages.size) {
                room.copy(messages = updatedMessages)
            } else {
                room
            }
        }
        saveChatRooms()
        println("💬 Deleted message: $messageId")
    }

    /**
     * Delete chat room
     */
    fun deleteChatRoom(peerId: String) {
        _chatRooms.value = _chatRooms.value.filter { it.peerId != peerId }
        saveChatRooms()
        println("💬 Deleted chat room: $peerId")
    }

    /**
     * Clear all chat rooms
     */
    fun clearAllChatRooms() {
        _chatRooms.value = emptyList()
        saveChatRooms()
        println("💬 Cleared all chat rooms")
    }

    /**
     * Total unread count
     */
    fun getTotalUnreadCount(): Int {
        return _chatRooms.value.sumOf { it.unreadCount }
    }

    /**
     * Add message to chat room
     */
    private fun addMessageToChatRoom(message: ChatMessage) {
        val peerId = if (message.senderId == myPeerId) {
            message.receiverId
        } else {
            message.senderId
        }

        val room = getChatRoom(peerId)
        if (room == null) {
            val peerDisplayName = message.senderDisplayName
                ?: message.receiverDisplayName
                ?: "Unknown"
            val newRoom = ChatRoom(
                peerId = peerId,
                peerDisplayName = peerDisplayName,
                messages = listOf(message)
            )
            _chatRooms.value = _chatRooms.value + newRoom
        } else {
            val updatedMessages = room.messages + message
            val updatedRoom = room.copy(messages = updatedMessages)
            updateChatRoom(updatedRoom)
        }

        saveChatRooms()
    }

    /**
     * Update a message
     */
    private fun updateMessage(message: ChatMessage) {
        _chatRooms.value = _chatRooms.value.map { room ->
            val updatedMessages = room.messages.map {
                if (it.id == message.id) message else it
            }
            if (updatedMessages != room.messages) {
                room.copy(messages = updatedMessages)
            } else {
                room
            }
        }
        saveChatRooms()
    }

    /**
     * Update chat room
     */
    private fun updateChatRoom(chatRoom: ChatRoom) {
        _chatRooms.value = _chatRooms.value.map {
            if (it.peerId == chatRoom.peerId) chatRoom else it
        }
        saveChatRooms()
    }

    /**
     * Send message via Nearby Connections
     */
    private fun sendMessageViaNearby(message: ChatMessage): Boolean {
        val manager = nearbyConnectionsManager
        if (manager == null) {
            println("⚠️ NearbyConnectionsManager not set")
            return false
        }

        return try {
            val json = gson.toJson(message)
            val data = json.toByteArray()
            manager.sendData(data, message.receiverId)
            true
        } catch (e: Exception) {
            println("❌ Failed to send message: ${e.message}")
            false
        }
    }

    /**
     * Load chat rooms from SharedPreferences
     */
    private fun loadChatRooms() {
        val json = prefs.getString("chat_rooms", null) ?: return
        try {
            val type = object : TypeToken<List<ChatRoom>>() {}.type
            _chatRooms.value = gson.fromJson(json, type)
        } catch (e: Exception) {
            println("❌ Failed to load chat rooms: ${e.message}")
        }
    }

    /**
     * Save chat rooms to SharedPreferences
     */
    private fun saveChatRooms() {
        val json = gson.toJson(_chatRooms.value)
        prefs.edit().putString("chat_rooms", json).apply()
    }

    /**
     * Cleanup
     */
    fun cleanup() {
        ephemeralMessageManager.cleanup()
    }
}

File: TrainBlink/ios/TrainBlink/Services/ChatManager.swift
//
//  ChatManager.swift
//  TrainBlink
//
//  Feature 5: 1-on-1 Chat Rooms
//  Manages chat rooms and message delivery
//
//  Created by TrainBlink Team
//  Copyright © 2025 TrainBlink. All rights reserved.
//

import Foundation
import Combine

/// Manages all chat rooms and message delivery
final class ChatManager: ObservableObject {

    // MARK: - Singleton

    static let shared = ChatManager()

    private init() {
        print("💬 ChatManager initialized")
        loadChatRooms()
        setupMessageExpiration()
    }

    // MARK: - Published Properties

    @Published private(set) var chatRooms: [ChatRoom] = []
    @Published private(set) var activeChatRoom: ChatRoom?

    // MARK: - Private Properties

    private let userDefaultsKey = "trainblink.chat_rooms"
    private let lock = NSLock()
    private var myPeerId: String = UIDevice.current.identifierForVendor?.uuidString ?? UUID().uuidString

    // MultipeerManager reference (for sending messages)
    private var multipeerManager: MultipeerManager?

    // Configuration
    private let maxMessagesPerRoom = 500  // Prevent unbounded growth
    private let messageTimeoutSeconds: TimeInterval = 30  // Message send timeout

    // MARK: - Public Methods

    /// Set my peer ID (should be called on app launch)
    func setMyPeerId(_ peerId: String) {
        lock.lock()
        defer { lock.unlock() }

        myPeerId = peerId
        print("💬 Set my peer ID: \(peerId)")
    }

    /// Set MultipeerManager reference (for message delivery)
    func setMultipeerManager(_ manager: MultipeerManager) {
        self.multipeerManager = manager
        print("💬 MultipeerManager connected")
    }

    /// Get or create a chat room with a peer
    /// - Parameter peer: The peer to chat with
    /// - Returns: The chat room
    func getChatRoom(with peer: Peer) -> ChatRoom {
        lock.lock()
        defer { lock.unlock() }

        // Check if chat room already exists
        if let existingRoom = chatRooms.first(where: { $0.peerId == peer.id }) {
            return existingRoom
        }

        // Create new chat room with encounter tracking
        let chatRoom = ChatRoom.createWithTracking(peer: peer)
        chatRooms.append(chatRoom)
        saveChatRooms()

        print("💬 Created chat room with \(peer.displayName)")

        // Log to Firebase Analytics
        AnalyticsManager.shared.logChatRoomCreated(
            peerId: peer.id,
            encounterCount: chatRoom.encounterCount
        )

        // Record encounter (Feature 8)
        EncounterTrackingManager.shared.recordEncounter(
            with: peer,
            interactionType: .chat
        )

        return chatRoom
    }

    /// Send a message to a peer
    /// - Parameters:
    ///   - text: Message text
    ///   - peerId: Recipient peer ID
    ///   - isEphemeral: Whether message is ephemeral
    /// - Returns: Result with the sent message
    func sendMessage(
        text: String,
        to peerId: String,
        isEphemeral: Bool = false
    ) -> Result<ChatMessage, ChatError> {
        lock.lock()
        defer { lock.unlock() }

        // Check if peer is blocked (Feature 7)
        if BlockingManager.shared.isBlocked(peerId: peerId) {
            print("🚫 Cannot send message to blocked peer: \(peerId)")
            return .failure(.peerBlocked)
        }

        // Find chat room
        guard let roomIndex = chatRooms.firstIndex(where: { $0.peerId == peerId }) else {
            return .failure(.chatRoomNotFound)
        }

        // Create message
        let message: ChatMessage
        if isEphemeral {
            message = ChatMessage.ephemeralMessage(
                text: text,
                from: myPeerId,
                to: peerId
            )
        } else {
            message = ChatMessage.textMessage(
                text: text,
                from: myPeerId,
                to: peerId
            )
        }

        // Add to chat room
        chatRooms[roomIndex].addMessage(message)
        saveChatRooms()

        // Track ephemeral message (Feature 6)
        if message.isEphemeral {
            EphemeralMessageManager.shared.trackMessage(message)
        }

        // Send via MultipeerConnectivity
        sendMessageData(message, to: peerId)

        print("💬 Sent message to \(peerId): \"\(text.prefix(20))...\"")

        // Log to Firebase Analytics
        AnalyticsManager.shared.logMessageSent(
            peerId: peerId,
            isEphemeral: isEphemeral,
            messageLength: text.count
        )

        return .success(message)
    }

    /// Handle received message data
    /// - Parameters:
    ///   - data: Message data
    ///   - senderId: Sender peer ID
    func handleReceivedMessage(data: Data, from senderId: String) {
        lock.lock()
        defer { lock.unlock() }

        // Check if sender is blocked (Feature 7)
        if BlockingManager.shared.isBlocked(peerId: senderId) {
            print("🚫 Rejecting message from blocked peer: \(senderId)")
            AnalyticsManager.shared.logMessageRejectedFromBlockedPeer(senderId: senderId)
            return
        }

        // Decode message
        guard let message = decodeMessage(data) else {
            print("❌ Failed to decode message from \(senderId)")
            return
        }

        // Find or create chat room
        let roomIndex: Int
        if let index = chatRooms.firstIndex(where: { $0.peerId == senderId }) {
            roomIndex = index
        } else {
            // Create new chat room for unknown peer
            let peer = Peer(id: senderId, displayName: "Peer \(senderId.prefix(8))")
            let chatRoom = ChatRoom.createWithTracking(peer: peer)
            chatRooms.append(chatRoom)
            roomIndex = chatRooms.count - 1

            // Log chat room created
            AnalyticsManager.shared.logChatRoomCreated(
                peerId: senderId,
                encounterCount: chatRoom.encounterCount
            )
        }

        // Mark as delivered and add to chat room
        var deliveredMessage = message
        deliveredMessage.markAsDelivered()
        chatRooms[roomIndex].addMessage(deliveredMessage)
        saveChatRooms()

        // Track ephemeral message (Feature 6)
        if deliveredMessage.isEphemeral {
            EphemeralMessageManager.shared.trackMessage(deliveredMessage)
        }

        print("💬 Received message from \(senderId): \"\(message.text.prefix(20))...\"")

        // Log to Firebase Analytics
        AnalyticsManager.shared.logMessageReceived(
            peerId: senderId,
            isEphemeral: message.isEphemeral,
            messageLength: message.text.count
        )

        // Record encounter (Feature 8)
        let peer = Peer(id: senderId, displayName: chatRooms[roomIndex].peerDisplayName)
        EncounterTrackingManager.shared.recordEncounter(
            with: peer,
            interactionType: .chat
        )
    }

    /// Mark all messages as read in a chat room
    /// - Parameter peerId: The peer ID
    func markAsRead(peerId: String) {
        lock.lock()
        defer { lock.unlock() }

        guard let index = chatRooms.firstIndex(where: { $0.peerId == peerId }) else {
            return
        }

        chatRooms[index].markAllAsRead(myId: myPeerId)
        saveChatRooms()

        print("💬 Marked all messages as read in chat with \(peerId)")
    }

    /// Close a chat room
    /// - Parameter peerId: The peer ID
    func closeChatRoom(peerId: String) {
        lock.lock()
        defer { lock.unlock() }

        guard let index = chatRooms.firstIndex(where: { $0.peerId == peerId }) else {
            return
        }

        chatRooms[index].close()
        saveChatRooms()

        print("💬 Closed chat room with \(peerId)")

        // Log to Firebase Analytics
        AnalyticsManager.shared.logChatRoomClosed(peerId: peerId)
    }

    /// Delete a chat room
    /// - Parameter peerId: The peer ID
    func deleteChatRoom(peerId: String) {
        lock.lock()
        defer { lock.unlock() }

        guard let index = chatRooms.firstIndex(where: { $0.peerId == peerId }) else {
            return
        }

        let removed = chatRooms.remove(at: index)
        saveChatRooms()

        print("💬 Deleted chat room with \(peerId)")

        // Log to Firebase Analytics
        AnalyticsManager.shared.logChatRoomDeleted(
            peerId: peerId,
            messageCount: removed.messageCount
        )
    }

    /// Get all active chat rooms (sorted by last message)
    var activeChatRooms: [ChatRoom] {
        lock.lock()
        defer { lock.unlock() }

        return chatRooms
            .filter { $0.isActive }
            .sorted { ($0.lastMessageAt ?? $0.createdAt) > ($1.lastMessageAt ?? $1.createdAt) }
    }

    /// Get total unread message count
    var totalUnreadCount: Int {
        lock.lock()
        defer { lock.unlock() }

        return chatRooms.reduce(0) { $0 + $1.unreadCount }
    }

    /// Clear all chat rooms (for testing or station exit)
    func clearAll() {
        lock.lock()
        defer { lock.unlock() }

        let count = chatRooms.count
        chatRooms.removeAll()
        saveChatRooms()

        print("💬 Cleared all chat rooms (\(count) rooms)")

        // Log to Firebase Analytics
        AnalyticsManager.shared.logAllChatRoomsCleared(count: count)
    }

    // MARK: - Private Methods

    /// Send message data via MultipeerConnectivity
    private func sendMessageData(_ message: ChatMessage, to peerId: String) {
        guard let manager = multipeerManager else {
            print("❌ MultipeerManager not set")
            return
        }

        guard let session = manager.getSession() else {
            print("❌ No active MultipeerConnectivity session")
            return
        }

        // Find peer in session
        let mcPeerId = session.connectedPeers.first(where: { $0.displayName == peerId })
        guard let peer = mcPeerId else {
            print("❌ Peer not connected: \(peerId)")
            return
        }

        // Encode message
        do {
            let data = try JSONEncoder().encode(message)

            // Send via MultipeerConnectivity
            try session.send(data, toPeers: [peer], with: .reliable)

            print("✅ Sent message data to \(peerId) (\(data.count) bytes)")
        } catch {
            print("❌ Failed to send message: \(error)")
        }
    }

    /// Decode message from data
    private func decodeMessage(_ data: Data) -> ChatMessage? {
        do {
            let message = try JSONDecoder().decode(ChatMessage.self, from: data)
            return message
        } catch {
            print("❌ Failed to decode message: \(error)")
            return nil
        }
    }

    /// Setup message expiration callback
    private func setupMessageExpiration() {
        EphemeralMessageManager.shared.onMessageExpired = { [weak self] messageId in
            self?.handleMessageExpired(messageId)
        }
    }

    /// Handle expired ephemeral message
    private func handleMessageExpired(_ messageId: String) {
        lock.lock()
        defer { lock.unlock() }

        // Find and remove message from all chat rooms
        for i in 0..<chatRooms.count {
            if let msgIndex = chatRooms[i].messages.firstIndex(where: { $0.id == messageId }) {
                let message = chatRooms[i].messages.remove(at: msgIndex)
                saveChatRooms()

                print("⏱️ Removed expired message from chat: \(message.text.prefix(20))...")
                break
            }
        }
    }

    /// Load chat rooms from UserDefaults
    private func loadChatRooms() {
        guard let data = UserDefaults.standard.data(forKey: userDefaultsKey) else {
            print("💬 No chat rooms found in UserDefaults")
            return
        }

        do {
            let decoded = try JSONDecoder().decode([ChatRoom].self, from: data)
            chatRooms = decoded
            print("💬 Loaded \(chatRooms.count) chat rooms")
        } catch {
            print("❌ Failed to decode chat rooms: \(error)")
        }
    }

    /// Save chat rooms to UserDefaults
    private func saveChatRooms() {
        do {
            let data = try JSONEncoder().encode(chatRooms)
            UserDefaults.standard.set(data, forKey: userDefaultsKey)
            print("💬 Saved \(chatRooms.count) chat rooms")
        } catch {
            print("❌ Failed to encode chat rooms: \(error)")
        }
    }
}

// MARK: - ChatError

/// Errors that can occur during chat operations
enum ChatError: Error, LocalizedError {
    case chatRoomNotFound
    case peerBlocked
    case sendFailed
    case receiveF

ailed
    case multipeerNotConnected
    case sessionNotActive

    var errorDescription: String? {
        switch self {
        case .chatRoomNotFound:
            return "Chat room not found"
        case .peerBlocked:
            return "This peer is blocked"
        case .sendFailed:
            return "Failed to send message"
        case .receiveFailed:
            return "Failed to receive message"
        case .multipeerNotConnected:
            return "MultipeerManager not connected"
        case .sessionNotActive:
            return "No active session"
        }
    }
}

// MARK: - Analytics Extensions

extension AnalyticsManager {

    /// Log chat room created
    func logChatRoomCreated(peerId: String, encounterCount: Int) {
        let parameters: [String: Any] = [
            "peer_id": peerId,
            "encounter_count": encounterCount
        ]
        logEvent("chat_room_created", parameters: parameters)
    }

    /// Log message sent
    func logMessageSent(peerId: String, isEphemeral: Bool, messageLength: Int) {
        let parameters: [String: Any] = [
            "peer_id": peerId,
            "is_ephemeral": isEphemeral,
            "message_length": messageLength
        ]
        logEvent("message_sent", parameters: parameters)
    }

    /// Log message received
    func logMessageReceived(peerId: String, isEphemeral: Bool, messageLength: Int) {
        let parameters: [String: Any] = [
            "peer_id": peerId,
            "is_ephemeral": isEphemeral,
            "message_length": messageLength
        ]
        logEvent("message_received", parameters: parameters)
    }

    /// Log message rejected from blocked peer
    func logMessageRejectedFromBlockedPeer(senderId: String) {
        let parameters: [String: Any] = [
            "sender_id": senderId
        ]
        logEvent("message_rejected_blocked_peer", parameters: parameters)
    }

    /// Log chat room closed
    func logChatRoomClosed(peerId: String) {
        let parameters: [String: Any] = [
            "peer_id": peerId
        ]
        logEvent("chat_room_closed", parameters: parameters)
    }

    /// Log chat room deleted
    func logChatRoomDeleted(peerId: String, messageCount: Int) {
        let parameters: [String: Any] = [
            "peer_id": peerId,
            "message_count": messageCount
        ]
        logEvent("chat_room_deleted", parameters: parameters)
    }

    /// Log all chat rooms cleared
    func logAllChatRoomsCleared(count: Int) {
        let parameters: [String: Any] = [
            "count": count
        ]
        logEvent("all_chat_rooms_cleared", parameters: parameters)
    }
}
